File: java/library/src/main/java/com/v2soft/styxlib/utils/Polls.java
package com.v2soft.styxlib.utils;

import com.v2soft.styxlib.messages.base.StyxTMessage;
import com.v2soft.styxlib.messages.base.StyxTMessageFID;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by vshcryabets on 12/10/14.
 */
public class Polls {
    protected Map<Integer, StyxTMessage> mMessagesMap;
    protected MessageTagPoll mTags;
    protected FIDPoll mFids;

    public Polls() {
        mFids = new FIDPoll();
        mTags = new MessageTagPoll();
        mMessagesMap = new HashMap<Integer, StyxTMessage>();
    }

    public FIDPoll getFIDPoll() {
        return mFids;
    }

    public MessageTagPoll getTagPoll() {
        return mTags;
    }

    public void releaseTag(int tag) {
        mMessagesMap.remove(tag);
        mTags.release(tag);
    }

    public void releaseFID(StyxTMessageFID message) {
        mFids.release(message.getFID());
    }

    public void putTMessage(int tag, StyxTMessage message) {
        mMessagesMap.put(tag, message);
    }

    // TODO probably we can use one call instead getTMessage && releaseTag
    public StyxTMessage getTMessage(int tag) {
        return mMessagesMap.get(tag);
    }
}

File: native/V2StyxLib/src/utils/Polls.cpp
/*
 * Polls.cpp
 *
 *  Created on: Dec 3, 2016
 *      Author: mrco
 */

#include "utils/Polls.h"

Polls::Polls() {
    mFids = new FIDPoll();
    mTags = new MessageTagPoll();
}

Polls::~Polls() {
	delete mFids;
	delete mTags;
	mMessagesMap.clear();
}

FIDPoll* Polls::getFIDPoll() {
	return mFids;
}

void Polls::releaseFID(StyxTMessageFID* message) {
	mFids->release(message->getFID());
}

void Polls::releaseTag(StyxTAG tag) {
	mMessagesMap.erase(tag);
	mTags->release(tag);
}

StyxTMessage* Polls::getTMessage(StyxTAG tag) {
	std::map<StyxTAG, StyxTMessage*>::iterator it = mMessagesMap.find(tag);
	if (it == mMessagesMap.end()) {
		return NULL;
	}
	return it->second;
}

MessageTagPoll* Polls::getTagPoll() {
	return mTags;
}

void Polls::putTMessage(StyxTAG tag, StyxTMessage* message) {
	mMessagesMap[tag] = message;
}
